packages/opentelemetry-instrumentation-watsonx/opentelemetry/instrumentation/watsonx/__init__.py

```diff
@@ -445,67 +445,66 @@ def _wrap(
 
     name = to_wrap.get("span_name")
 
-    span = tracer.start_span(
+    with tracer.start_span(
         name,
         kind=SpanKind.CLIENT,
         attributes={
             SpanAttributes.LLM_SYSTEM: "Watsonx",
             SpanAttributes.LLM_REQUEST_TYPE: LLMRequestTypeValues.COMPLETION.value,
         },
-    )
-
-    _set_api_attributes(span)
-    if "generate" in name:
-        _set_input_attributes(span, instance, kwargs)
-        if to_wrap.get("method") == "generate_text_stream":
-            if (raw_flag := kwargs.get("raw_response", None)) is None:
-                kwargs = {**kwargs, "raw_response": True}
-            elif raw_flag is False:
-                kwargs["raw_response"] = True
-
-    try:
-        start_time = time.time()
-        response = wrapped(*args, **kwargs)
-        end_time = time.time()
-    except Exception as e:
-        end_time = time.time()
-        duration = end_time - start_time if "start_time" in locals() else 0
-
-        attributes = {
-            "error.type": e.__class__.__name__,
-        }
+    ) as span:
+
+        _set_api_attributes(span)
+        if "generate" in name:
+            _set_input_attributes(span, instance, kwargs)
+            if to_wrap.get("method") == "generate_text_stream":
+                if (raw_flag := kwargs.get("raw_response", None)) is None:
+                    kwargs = {**kwargs, "raw_response": True}
+                elif raw_flag is False:
+                    kwargs["raw_response"] = True
+
+        try:
+            start_time = time.time()
+            response = wrapped(*args, **kwargs)
+            end_time = time.time()
+        except Exception as e:
+            end_time = time.time()
+            duration = end_time - start_time if "start_time" in locals() else 0
+
+            attributes = {
+                "error.type": e.__class__.__name__,
+            }
 
-        if duration > 0 and duration_histogram:
-            duration_histogram.record(duration, attributes=attributes)
-        if exception_counter:
-            exception_counter.add(1, attributes=attributes)
+            if duration > 0 and duration_histogram:
+                duration_histogram.record(duration, attributes=attributes)
+            if exception_counter:
+                exception_counter.add(1, attributes=attributes)
 
-        raise e
+            raise e
 
-    if "generate" in name:
```

Comment:
I think this won't work well with the generator response. You need to manually close the span on all cases

Reply:
Thanks, I will investigate and update with more details.

Reply:
Explicitly call span.__exit__ for exception, just like using `with` statement to do it automatically.

Reply:
@nirga please review again. `span.__exit__` is the default operation to record exception information and call span.end(), and no impact to generator response.

Reply:
<img width="1013" alt="image" src="https://github.com/user-attachments/assets/028d7100-23e9-4f0f-8b49-fc9577e9e64b">

Reply:
But looks like it's not ideal because of the parent span for workflow is missing. Do you have any suggestion?

Reply:
@jinsongo why not just call `span.end()`?

Reply:
@nirga __eixt__ will call span.end() and also handle the exception.

Reply:
BTW, bedrock instrumentation use `with` statement and no 'Generator' problem, so the exception can be handled by `span.__exit__` automatically when exception raised from `line 166 response = fn(*args, **kwargs)`
https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-bedrock/opentelemetry/instrumentation/bedrock/__init__.py#L163-L171

Reply:
`span.__exit__` implementation:
https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L986-L1008
`span.__exit__` of parent class:
https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/trace/span.py#L186-L194

Reply:
Yes @jinsongo but you're setting the exception yourself anyway so you don't really need this, right?

Reply:
I prefer to set the exception information into the span by simply calling `__exit__`.
As you know, without the fix, I cannot see any span sent after the exception raised.

Reply:
@jinsongo why? You didn't call `__enter__` so I don't think it's preferred on explicitly logging the exception on the span